navigation/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/camera/NavigationCamera.java

```diff
@@ -0,0 +1,195 @@
+  private static int CAMERA_ZOOM = 17;
+
+  private MapboxMap mapboxMap;
+  private MapboxNavigation navigation;
+
+  private Location location;
+  private double targetDistance;
+  private boolean trackingEnabled = true;
+
+  public NavigationCamera(@NonNull Context context, @NonNull MapboxMap mapboxMap,
+                          @NonNull MapboxNavigation navigation) {
+    this.mapboxMap = mapboxMap;
+    this.navigation = navigation;
+    initialize(context);
+  }
+
+  public void start(DirectionsRoute route) {
+    if (route != null) {
+      CameraPosition cameraPosition = buildCameraPositionFromRoute(route);
+      mapboxMap.animateCamera(CameraUpdateFactory.newCameraPosition(cameraPosition), 2000,
+        new MapboxMap.CancelableCallback() {
+          @Override
+          public void onCancel() {
+
+          }
+
+          @Override
+          public void onFinish() {
+            navigation.addProgressChangeListener(NavigationCamera.this);
+          }
+        });
+    } else {
+      navigation.addProgressChangeListener(NavigationCamera.this);
+    }
+  }
+
+  public void resume(Location location) {
+    if (location != null) {
+      CameraPosition position = buildCameraPositionFromLocation(location);
+      mapboxMap.animateCamera(CameraUpdateFactory.newCameraPosition(position), 2000,
+        new MapboxMap.CancelableCallback() {
+          @Override
+          public void onCancel() {
+
+          }
+
+          @Override
+          public void onFinish() {
+            navigation.addProgressChangeListener(NavigationCamera.this);
+          }
+        });
+    } else {
+      navigation.addProgressChangeListener(NavigationCamera.this);
+    }
+  }
+
+  @Override
+  public void onProgressChange(Location location, RouteProgress routeProgress) {
+    this.location = location;
+    easeCameraToLocation(location);
+  }
+
+  public void setCameraTrackingLocation(boolean trackingEnabled) {
+    this.trackingEnabled = trackingEnabled;
+  }
+
+  public void resetCameraPosition() {
+    this.trackingEnabled = true;
+    if (location != null) {
+      LatLng target = new LatLng(location.getLatitude(), location.getLongitude());
+      CameraPosition cameraPosition = new CameraPosition.Builder()
+        .tilt(CAMERA_TILT)
+        .zoom(CAMERA_ZOOM)
+        .target(target)
+        .bearing(location.getBearing())
+        .build();
+      mapboxMap.easeCamera(CameraUpdateFactory.newCameraPosition(cameraPosition), 750, false);
+    }
+  }
+
+  private void easeCameraToLocation(Location location) {
+    CameraPosition cameraPosition = buildCameraPositionFromLocation(location);
+
+    if (trackingEnabled) {
+      mapboxMap.easeCamera(CameraUpdateFactory.newCameraPosition(cameraPosition), 1000, false);
+    }
+  }
+
+  @NonNull
+  private CameraPosition buildCameraPositionFromRoute(DirectionsRoute route) {
+    LineString lineString = LineString.fromPolyline(route.getGeometry(), Constants.PRECISION_6);
```

Comment:
is there a better solution then decoding the entire route here? Possible to just use step or to use a different way to get camera position?